pyomo/contrib/solver/solvers/gurobi_direct_minlp.py

```diff
@@ -0,0 +1,655 @@
+_LINEAR = ExprType.LINEAR
+_QUADRATIC = ExprType.QUADRATIC
+_VARIABLE = ExprType.VARIABLE
+
+_function_map = {}
+_domain_map = ComponentMap()
+
+gurobipy, gurobipy_available = attempt_import('gurobipy', minimum_version='12.0.0')
+if gurobipy_available:
+    from gurobipy import GRB, nlfunc
+
+    _function_map.update(
+        {
+            'exp': (_GENERAL, nlfunc.exp),
+            'log': (_GENERAL, nlfunc.log),
+            'log10': (_GENERAL, nlfunc.log10),
+            'sin': (_GENERAL, nlfunc.sin),
+            'cos': (_GENERAL, nlfunc.cos),
+            'tan': (_GENERAL, nlfunc.tan),
+            'sqrt': (_GENERAL, nlfunc.sqrt),
+            # Not supporting any of these right now--we'd have to build them from the
+            # above:
+            # 'asin': None,
+            # 'sinh': None,
+            # 'asinh': None,
+            # 'acos': None,
+            # 'cosh': None,
+            # 'acosh': None,
+            # 'atan': None,
+            # 'tanh': None,
+            # 'atanh': None,
+            # 'ceil': None,
+            # 'floor': None,
+        }
+    )
+
+    _domain_map.update(
+        (
+            (Binary, (GRB.BINARY, -float('inf'), float('inf'))),
+            (Integers, (GRB.INTEGER, -float('inf'), float('inf'))),
+            (NonNegativeIntegers, (GRB.INTEGER, 0, float('inf'))),
+            (NonPositiveIntegers, (GRB.INTEGER, -float('inf'), 0)),
+            (NonNegativeReals, (GRB.CONTINUOUS, 0, float('inf'))),
+            (NonPositiveReals, (GRB.CONTINUOUS, -float('inf'), 0)),
+            (Reals, (GRB.CONTINUOUS, -float('inf'), float('inf'))),
+        )
+    )
+
+
+"""
+In Gurobi 12:
+
+If you have f(x) == 0, you must write it as z == f(x) and then write z == 0.
+Basically, you must introduce auxiliary variables for all the general nonlinear
+parts. (And no worries about additively separable or anything--they do that 
+under the hood).
+
+In this implementation, we replace the *entire* LHS of the constraint with the
+auxiliary variable rather than just the nonlinear part. Otherwise we would really
+need to keep track of what nonlinear subexpressions we had already replaced and make
+sure to use the same auxiliary variables, and from what we know, this is probably not
+worth it.
+
+We are not using Gurobi's '.nl' attribute at all for now--its usage seems like the
+exception rather than the rule, so we will let Gurobi expand the expressions for now.
+"""
+
+
+def _create_grb_var(visitor, pyomo_var, name=""):
+    pyo_domain = pyomo_var.domain
+    if pyo_domain in _domain_map:
+        domain, domain_lb, domain_ub = _domain_map[pyo_domain]
+    else:
+        raise ValueError(
+            "Unsupported domain for Var '%s': %s" % (pyomo_var.name, pyo_domain)
+        )
+    lb = max(domain_lb, pyomo_var.lb) if pyomo_var.lb is not None else domain_lb
+    ub = min(domain_ub, pyomo_var.ub) if pyomo_var.ub is not None else domain_ub
+    return visitor.grb_model.addVar(lb=lb, ub=ub, vtype=domain, name=name)
```

Comment:
Additional note: we should really consider providing an API that returns the bounds as a float (i.e., doesn't return `None`)